components/org.wso2.carbon.identity.notification.push.common/src/main/java/org/wso2/carbon/identity/notification/push/common/PushChallengeValidator.java

```diff
@@ -0,0 +1,227 @@
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.wso2.carbon.identity.notification.push.common;
+
+import com.nimbusds.jose.JOSEException;
+import com.nimbusds.jose.JWSVerifier;
+import com.nimbusds.jose.crypto.RSASSAVerifier;
+import com.nimbusds.jwt.JWTClaimsSet;
+import com.nimbusds.jwt.SignedJWT;
+import org.apache.commons.lang.StringUtils;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.wso2.carbon.identity.notification.push.common.exception.PushTokenValidationException;
+
+import java.security.KeyFactory;
+import java.security.NoSuchAlgorithmException;
+import java.security.interfaces.RSAPublicKey;
+import java.security.spec.InvalidKeySpecException;
+import java.security.spec.X509EncodedKeySpec;
+import java.text.ParseException;
+import java.util.Base64;
+import java.util.Date;
+
+/**
+ * JWT token validator for Push notification scenarios.
+ */
+public class PushChallengeValidator {
+
+    private static final String DOT_SEPARATOR = ".";
+    public static final String SIGNING_ALGORITHM = "RSA";
+    private static final long DEFAULT_TIME_SKEW_IN_SECONDS = 300L;
+    private static final Log log = LogFactory.getLog(PushChallengeValidator.class);
+
+    /**
+     * Validate the JWT token.
+     *
+     * @param jwt       JWT token to be validated
+     * @param publicKey Public key used for signing the JWT
+     * @return JWTClaimsSet
+     * @throws PushTokenValidationException Error when validating the JWT token
+     */
+    public static JWTClaimsSet getValidatedClaimSet(String jwt, String publicKey) throws PushTokenValidationException {
+
+        if (!isJWT(jwt)) {
+            throw new PushTokenValidationException("Token is not a valid JWT.");
+        }
+
+        try {
+            SignedJWT signedJWT = SignedJWT.parse(jwt);
+            JWTClaimsSet claimsSet = signedJWT.getJWTClaimsSet();
+            if (claimsSet == null) {
+                throw new PushTokenValidationException("Token validation failed. Claim values were not found.");
+            }
+
+            if (!validateSignature(publicKey, signedJWT)) {
+                throw new PushTokenValidationException("Token signature validation failed.");
+            }
+
+            if (!checkExpirationTime(claimsSet.getExpirationTime())) {
+                throw new PushTokenValidationException("Token validation failed. JWT is expired.");
+            }
+            if (!checkNotBeforeTime(claimsSet.getNotBeforeTime())) {
```

Comment:
nbf is not required. Hence remove the validation and remove from the token as well.